docs/proposals/20220910-enhancement-of-servicetopology.md

```diff
@@ -0,0 +1,213 @@
+
+## Summary
+
+Servicetopology filter in Yurthub can work properly when endpointslice or endpoints resource change. But servicetopology filter can't work properly when nodePool or service resource change, like updating topology annotation for service or removing node from nodePool, endpointslice or endpoints on the edge nodes will not be updated.
+
+## Motivation
+A new edge node may add to an existed nodepool, or an old edge node may remove from the nodepool. And users may modify the topology configuration in service annotations, like removing `openyurt.io/topologyKeys` or changing topology configuration from `openyurt.io/nodepool` to `kubernetes.io/hostname`.
+If endpointslice or endpoints can be changed along with nodepool or service, the servicetopology filter can work properly when nodePool or service resource change. According to the latest nodepool, service access traffic can circulate to new edge nodes and not to old edge nodes. And according to the latest service, the servicetopology filter in YurtHub can change the type of servicetopology or disable servicetopology, etc.
+
+![nodepoolchange](../img/service-topology-nodepool-change.png)
+
+![servicechange](../img/service-topology-service-change.png)
+
+### Goals
+
+- Make servicetopology filter in Yurthub work properly when service change
+- Make servicetopology filter in Yurthub work properly when nodepool change
+
+## Proposal
+
+To make servicetopology filter in Yurthub work properly when service or nodepool change, we need two controllers, one for endpoints and another for endpointslice.
+
+### Endpoints controller
+The endpoints contoller will watch the change of service and nodepool, the event handlers will enqueue the necessary endpoints to the workqueue of controller, then the controller can modify the trigger annotation `openyurt.io/updateTrigger` for the endpoints. The value of trigger annotation is a timestamp, when the annotation of endpoints is modified, then the servicetopology filter can sense the change of endpoints and will get the latest service and nodepool when filtering.
+
+#### ·Service event handler
+
+When the servicetopology configuration in service.Annotations is modified, the handler will enqueue the endpoints of that service.
+```go
+func (e *EnqueueEndpointsForService)Update(evt event.UpdateEvent,
+	q workqueue.RateLimitingInterface) {
+	...
+
+	if newSvc.Annotations[AnnotationServiceTopologyKey] == oldSvc.Annotations[AnnotationServiceTopologyKey] {
+		return
+	}
+	nsName := types.NamespacedName{Namespace: newSvc.Namespace, Name: newSvc.Name}
+	q.Add(reconcile.Request{NamespacedName: nsName})
+	return
+}
+```
+
+#### ·Nodepool event handler
+In order to reduce the traffic which is generated by the change of endpoints, nodepool event handler will only enqueue those endpoints which are affected by nodepool changes. First, the servicetopology type of endpoints must be nodepool type. Then the endpoints that need to be enqueued must met the following conditions:
```

Comment:
I think it's more reasonable to list/watch Node resource instead of NodePool resource, so added `endpoints` and `endpointslice` controller will not depend on the `yurt-app-manager` component.

Reply:
@LinFCai Thanks for your reply. I have got your consideration for enqueuing all endpoints/endpointslices of NodePool. and i agree with you. please skip this comment.